heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java

```diff
@@ -231,36 +226,135 @@ public void close() {
    */
   private Map<Integer, List<InstanceId>> getFFDAllocation() {
     Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
-    return assignInstancesToContainers(parallelismMap, 1, 1);
+    HashMap<Integer, List<InstanceId>> allocation = new HashMap<>();
+    assignInstancesToContainers(new ArrayList<Container>(), allocation, parallelismMap, 1);
+    return allocation;
   }
 
   /**
    * Get the instances' allocation based on the First Fit Decreasing algorithm
    *
    * @return Map &lt; containerId, list of InstanceId belonging to this container &gt;
    */
+
   private Map<Integer, List<InstanceId>> getFFDAllocation(PackingPlan currentPackingPlan,
                                                           Map<String, Integer> componentChanges) {
-    int numContainers = currentPackingPlan.getContainers().size();
+    Map<String, Integer> componentsToScaleDown = getComponentsToScaleDown(componentChanges);
+    Map<String, Integer> componentsToScaleUp = getComponentsToScaleUp(componentChanges);
+
+    ArrayList<Container> containers = getContainers(currentPackingPlan);
+    Map<Integer, List<InstanceId>> allocation = getAllocation(currentPackingPlan);
     int maxInstanceIndex = 0;
     for (PackingPlan.ContainerPlan containerPlan : currentPackingPlan.getContainers()) {
       for (PackingPlan.InstancePlan instancePlan : containerPlan.getInstances()) {
         maxInstanceIndex = Math.max(maxInstanceIndex, instancePlan.getTaskId());
       }
     }
-    return assignInstancesToContainers(componentChanges, numContainers + 1, maxInstanceIndex + 1);
+    if (!componentsToScaleDown.isEmpty()) {
+      removeInstancesFromContainers(containers, allocation, componentsToScaleDown);
+    }
+    if (!componentsToScaleUp.isEmpty()) {
+      assignInstancesToContainers(containers, allocation, componentsToScaleUp,
+          maxInstanceIndex + 1);
+    }
+    cleanAllocation(allocation);
+    return allocation;
+  }
+
+  /**
+   * Removes containers from tha allocation that do not contain any instances
+   */
+  private void cleanAllocation(Map<Integer, List<InstanceId>> allocation) {
+    Iterator<Map.Entry<Integer, List<InstanceId>>> entries = allocation.entrySet().iterator();
```

Comment:
Instead of iterating over entries it's more idiomatic (and reads easier IMO since it's always obvious what the key and value represents by their naming) to iterate over keys:
```
Iterator<Integer> containerIds = allocation.keySet().iterator();
while (containerIds.hasNext() {
  Integer containerId = containerIds.next();
  if (allocation.get(containerId) == null)
   allocation.remove(containerId);
...
```
Or even better using `for (Integer containerId : allocation.keySet())` if that doesn't cause issues during deletion.